lib/rbs/definition_builder.rb

```diff
@@ -786,6 +786,59 @@ def expand_alias(type_name)
       entry.decl.type
     end
 
+    def type_alias_cache
+      require 'tsort'
+      # Initialize hash(a directed graph) with an empty array
+      cache = env.alias_decls.transform_values { |_| [] }
+      # Iterate over alias declarations inserted into environment
+      env.alias_decls.each do |name, entry|
+        extract_type_aliases(entry.decl.type, cache[name])
+      end
+      # Create a proc for iterating over nodes
+      each_node = lambda { |&block| cache.each_key(&block) }
+      # Create a proc for iterating over children of a node
+      each_child = lambda do |name, &block|
+        if cache[name].nil?
+          # nop
+        else
+          cache[name].each(&block)
+        end
+      end
+      # Build a hash table of type names & strongly connected aliases index
+      scc_hash = {}
+      # Apply TSort technique & fetch strongly connected components
+      strongly_connected_components = TSort.strongly_connected_components(each_node, each_child)
```

Comment:
I don't think building strongly connected components is necessary.
We can build the transitive closure of dependencies from type name. And test if the left hand side is included in the closure. I believe the definition with transitive closure is more straightforward.

1. Define _dependent type names_ for given type `T` (`DD(T)`) same as `extract_type_aliases`.
2. Calculate transitive closure of _dependent type names_ (`D(T)`) of one type.
3. Test if the type alias definition `type t = S` includes a circular reference by `D(S) ∋ t`.

On the other hand, I'm not very sure if which one is faster.